clippy_lints/src/methods/swap_with_temporary.rs

```diff
@@ -0,0 +1,128 @@
+    if let ExprKind::Path(QPath::Resolved(_, func_path)) = func.kind
+        && let Some(func_def_id) = func_path.res.opt_def_id()
+        && cx.tcx.is_diagnostic_item(sym::mem_swap, func_def_id)
+    {
+        match (arg_kind(&args[0]), arg_kind(&args[1])) {
+            (ArgKind::RefMutToTemp(left_temp), ArgKind::RefMutToTemp(right_temp)) => {
+                emit_lint_useless(cx, expr, func, &args[0], &args[1], left_temp, right_temp);
+            },
+            (ArgKind::RefMutToTemp(left_temp), right) => emit_lint_assign(cx, expr, &right, left_temp),
+            (left, ArgKind::RefMutToTemp(right_temp)) => emit_lint_assign(cx, expr, &left, right_temp),
+            _ => {},
+        }
+    }
+}
+
+enum ArgKind<'tcx> {
+    PlaceExpr(&'tcx Expr<'tcx>),
+    RefMutToPlace(&'tcx Expr<'tcx>),
+    RefMutToTemp(&'tcx Expr<'tcx>),
+}
+
+fn arg_kind<'tcx>(arg: &'tcx Expr<'tcx>) -> ArgKind<'tcx> {
+    if let ExprKind::AddrOf(BorrowKind::Ref, _, target) = arg.kind {
+        if target.is_syntactic_place_expr() {
+            ArgKind::RefMutToPlace(target)
+        } else {
+            ArgKind::RefMutToTemp(target)
+        }
+    } else {
+        ArgKind::PlaceExpr(arg)
+    }
+}
+
+fn emit_lint_useless(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    func: &Expr<'_>,
+    left: &Expr<'_>,
+    right: &Expr<'_>,
+    left_temp: &Expr<'_>,
+    right_temp: &Expr<'_>,
+) {
+    span_lint_and_then(
+        cx,
+        SWAP_WITH_TEMPORARY,
+        expr.span,
+        "swapping temporary values has no effect",
+        |diag| {
+            const DROP_MSG: &str = "drop them if creating these temporary expressions is necessary";
+
+            diag.span_note(left_temp.span, MSG_TEMPORARY);
+            diag.span_note(right_temp.span, MSG_TEMPORARY);
+
+            // If the `swap()` is a statement by itself, just propose to replace `swap(&mut a, &mut b)` by `a;
+            // b;` in order to drop `a` and `b` while acknowledging their side effects. If the
+            // `swap()` call is part of a larger expression, replace it by `{core,
+            // std}::mem::drop((a, b))`.
+            if matches!(cx.tcx.parent_hir_node(expr.hir_id), Node::Stmt(..)) {
+                diag.multipart_suggestion(
+                    DROP_MSG,
+                    vec![
+                        (func.span.with_hi(left_temp.span.lo()), String::new()),
+                        (left_temp.span.between(right_temp.span), String::from("; ")),
+                        (expr.span.with_lo(right_temp.span.hi()), String::new()),
+                    ],
+                    Applicability::MachineApplicable,
+                );
+            } else {
+                diag.multipart_suggestion(
+                    DROP_MSG,
+                    vec![
+                        (
+                            func.span,
+                            format!("{}::mem::drop(", if is_no_std_crate(cx) { "core" } else { "std" }),
```

Comment:
I think this part could be simpler written by using `clippy_utils::std_or_core`.
```suggestion
                            format!("{}::mem::drop(", std_or_core(cx)),
```